File: tests/rosetta/transpiler/FS/24-game-solve.fs
// Generated 2025-07-24 07:28 +0700

exception Return

let mutable _nowSeed = 0
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525 + 1013904223) % 2147483647
        _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
type Rational = {
    num: int
    denom: int
}
type Expr =
    | Num of Rational
    | Bin of int * Expr * Expr
open System

let OP_ADD: int = 1
let OP_SUB: int = 2
let OP_MUL: int = 3
let OP_DIV: int = 4
let rec binEval (op: int) (l: Expr) (r: Expr) =
    let mutable __ret : Rational = Unchecked.defaultof<Rational>
    let mutable op = op
    let mutable l = l
    let mutable r = r
    try
        let lv = exprEval l
        let rv = exprEval r
        if op = OP_ADD then
            __ret <- { num = ((lv.num) * (rv.denom)) + ((lv.denom) * (rv.num)); denom = (lv.denom) * (rv.denom) }
            raise Return
        if op = OP_SUB then
            __ret <- { num = ((lv.num) * (rv.denom)) - ((lv.denom) * (rv.num)); denom = (lv.denom) * (rv.denom) }
            raise Return
        if op = OP_MUL then
            __ret <- { num = (lv.num) * (rv.num); denom = (lv.denom) * (rv.denom) }
            raise Return
        __ret <- { num = (lv.num) * (rv.denom); denom = (lv.denom) * (rv.num) }
        raise Return
        __ret
    with
        | Return -> __ret
and binString (op: int) (l: Expr) (r: Expr) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable op = op
    let mutable l = l
    let mutable r = r
    try
        let ls = exprString l
        let rs = exprString r
        let mutable opstr: string = ""
        if op = OP_ADD then
            opstr <- " + "
        else
            if op = OP_SUB then
                opstr <- " - "
            else
                if op = OP_MUL then
                    opstr <- " * "
                else
                    opstr <- " / "
        __ret <- ((("(" + ls) + opstr) + rs) + ")"
        raise Return
        __ret
    with
        | Return -> __ret
and newNum (n: int) =
    let mutable __ret : Expr = Unchecked.defaultof<Expr>
    let mutable n = n
    try
        __ret <- Num({ num = n; denom = 1 })
        raise Return
        __ret
    with
        | Return -> __ret
and exprEval (x: Expr) =
    let mutable __ret : Rational = Unchecked.defaultof<Rational>
    let mutable x = x
    try
        __ret <- (match x with
            | Num(v) -> v
            | Bin(op, l, r) -> binEval op l r)
        raise Return
        __ret
    with
        | Return -> __ret
and exprString (x: Expr) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable x = x
    try
        __ret <- (match x with
            | Num(v) -> string (v.num)
            | Bin(op, l, r) -> binString op l r)
        raise Return
        __ret
    with
        | Return -> __ret
let n_cards: int = 4
let goal: int = 24
let digit_range: int = 9
let rec solve (xs: Expr array) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable xs = xs
    try
        if (Seq.length xs) = 1 then
            let f = exprEval (xs.[0])
            if ((f.denom) <> 0) && ((f.num) = ((f.denom) * goal)) then
                printfn "%s" (string (exprString (xs.[0])))
                __ret <- true
                raise Return
            __ret <- false
            raise Return
        let mutable i: int = 0
        while i < (Seq.length xs) do
            let mutable j: int = i + 1
            while j < (Seq.length xs) do
                let mutable rest = [||]
                let mutable k: int = 0
                while k < (Seq.length xs) do
                    if (k <> i) && (k <> j) then
                        rest <- Array.append rest [|xs.[k]|]
                    k <- k + 1
                let a = xs.[i]
                let b = xs.[j]
                for op in [|OP_ADD; OP_SUB; OP_MUL; OP_DIV|] do
                    let mutable node: Expr = Bin(op, a, b)
                    if solve (Array.append rest [|node|]) then
                        __ret <- true
                        raise Return
                let mutable node: Expr = Bin(OP_SUB, b, a)
                if solve (Array.append rest [|node|]) then
                    __ret <- true
                    raise Return
                node <- Bin(OP_DIV, b, a)
                if solve (Array.append rest [|node|]) then
                    __ret <- true
                    raise Return
                j <- j + 1
            i <- i + 1
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    try
        let mutable iter: int = 0
        while iter < 10 do
            let mutable cards = [||]
            let mutable i: int = 0
            while i < n_cards do
                let n = ((_now()) % (digit_range - 1)) + 1
                cards <- Array.append cards [|newNum n|]
                printfn "%s" (" " + (string n))
                i <- i + 1
            printfn "%s" ":  "
            if not (solve cards) then
                printfn "%s" "No solution"
            iter <- iter + 1
        __ret
    with
        | Return -> __ret
main()

File: tests/rosetta/transpiler/Pascal/24-game-solve.pas
{$mode objfpc}
program Main;
uses SysUtils;
type ExprArray = array of Expr;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
type Rational = record
  num: integer;
  denom: integer;
end;
type Expr = record
end;
var
  OP_ADD: integer;
  OP_SUB: integer;
  OP_MUL: integer;
  OP_DIV: integer;
  binEval_lv: integer;
  binEval_rv: integer;
  binString_ls: integer;
  binString_rs: integer;
  binString_opstr: string;
  n_cards: integer;
  goal: integer;
  digit_range: integer;
  solve_f: Rational;
  solve_i: integer;
  solve_j: integer;
  solve_rest: array of Expr;
  solve_k: integer;
  solve_a: Expr;
  solve_b: Expr;
  solve_node: integer;
  op: integer;
  main_iter: integer;
  main_cards: array of Expr;
  main_i: integer;
  main_n: integer;
function makeExpr(): Expr;
begin
end;
function makeRational(num: integer; denom: integer): Rational;
begin
  Result.num := num;
  Result.denom := denom;
end;
function binEval(op: integer; l: Expr; r: Expr): Rational;
begin
  binEval_lv := exprEval(l);
  binEval_rv := exprEval(r);
  if op = OP_ADD then begin
  exit(makeRational((binEval_lv.num * binEval_rv.denom) + (binEval_lv.denom * binEval_rv.num), binEval_lv.denom * binEval_rv.denom));
end;
  if op = OP_SUB then begin
  exit(makeRational((binEval_lv.num * binEval_rv.denom) - (binEval_lv.denom * binEval_rv.num), binEval_lv.denom * binEval_rv.denom));
end;
  if op = OP_MUL then begin
  exit(makeRational(binEval_lv.num * binEval_rv.num, binEval_lv.denom * binEval_rv.denom));
end;
  exit(makeRational(binEval_lv.num * binEval_rv.denom, binEval_lv.denom * binEval_rv.num));
end;
function binString(op: integer; l: Expr; r: Expr): string;
begin
  binString_ls := exprString(l);
  binString_rs := exprString(r);
  binString_opstr := '';
  if op = OP_ADD then begin
  binString_opstr := ' + ';
end else begin
  if op = OP_SUB then begin
  binString_opstr := ' - ';
end else begin
  if op = OP_MUL then begin
  binString_opstr := ' * ';
end else begin
  binString_opstr := ' / ';
end;
end;
end;
  exit(((('(' + binString_ls) + binString_opstr) + binString_rs) + ')');
end;
function newNum(n: integer): Expr;
begin
  exit(makeNum(makeRational(n, 1)));
end;
function exprEval(x: Expr): Rational;
begin
  exit(IfThen(x = Num(v), v, binEval(op, l, r)));
end;
function exprString(x: Expr): string;
begin
  exit(IfThen(x = Num(v), IntToStr(v.num), binString(op, l, r)));
end;
function solve(xs: ExprArray): boolean;
begin
  if Length(xs) = 1 then begin
  solve_f := exprEval(xs[0]);
  if (solve_f.denom <> 0) and (solve_f.num = (solve_f.denom * goal)) then begin
  writeln(exprString(xs[0]));
  exit(true);
end;
  exit(false);
end;
  solve_i := 0;
  while solve_i < Length(xs) do begin
  solve_j := solve_i + 1;
  while solve_j < Length(xs) do begin
  solve_rest := [];
  solve_k := 0;
  while solve_k < Length(xs) do begin
  if (solve_k <> solve_i) and (solve_k <> solve_j) then begin
  solve_rest := concat(solve_rest, [xs[solve_k]]);
end;
  solve_k := solve_k + 1;
end;
  solve_a := xs[solve_i];
  solve_b := xs[solve_j];
  solve_node := makeBin(OP_ADD, solve_a, solve_b);
  for op in [OP_ADD, OP_SUB, OP_MUL, OP_DIV] do begin
  solve_node := makeBin(op, solve_a, solve_b);
  if solve(concat(solve_rest, [solve_node])) then begin
  exit(true);
end;
end;
  solve_node := makeBin(OP_SUB, solve_b, solve_a);
  if solve(concat(solve_rest, [solve_node])) then begin
  exit(true);
end;
  solve_node := makeBin(OP_DIV, solve_b, solve_a);
  if solve(concat(solve_rest, [solve_node])) then begin
  exit(true);
end;
  solve_j := solve_j + 1;
end;
  solve_i := solve_i + 1;
end;
  exit(false);
end;
procedure main();
begin
  main_iter := 0;
  while main_iter < 10 do begin
  main_cards := [];
  main_i := 0;
  while main_i < n_cards do begin
  main_n := (_now() mod (digit_range - 1)) + 1;
  main_cards := concat(main_cards, [newNum(main_n)]);
  writeln(' ' + IntToStr(main_n));
  main_i := main_i + 1;
end;
  writeln(':  ');
  if not solve(main_cards) then begin
  writeln('No solution');
end;
  main_iter := main_iter + 1;
end;
end;
begin
  init_now();
  OP_ADD := 1;
  OP_SUB := 2;
  OP_MUL := 3;
  OP_DIV := 4;
  n_cards := 4;
  goal := 24;
  digit_range := 9;
  main();
end.
